modules/frontend/queue/queue.go

```diff
@@ -197,13 +190,22 @@ FindQueue:
 	goto FindQueue
 }
 
-func (q *RequestQueue) forgetDisconnectedQueriers(_ context.Context) error {
+func (q *RequestQueue) cleanupQueues(_ context.Context) error {
 	q.mtx.Lock()
 	defer q.mtx.Unlock()
 
-	if q.queues.forgetDisconnectedQueriers(time.Now()) > 0 {
-		// We need to notify goroutines cause having removed some queriers
-		// may have caused a resharding.
+	removedQueue := false
+
+	// look for 0 len queues and remove them
+	for userID, uq := range q.queues.userQueues {
```

Comment:
We should move this logic to user_queues as before, I think it's the responsibility of user_queues to know how to cleanup queues:

```go
if q.queues.deleteEmptyQueues(){
   q.cond.Broadcast()
}
```

Reply:
queue and user_queues are pretty grossly intertwined, but you're not wrong. fixing!